java/client/src/main/java/glide/managers/CommandManager.java

```diff
@@ -31,343 +29,860 @@
+            return errorFuture;
+        }
+
+        try {
+            // Serialize the protobuf command request
+            byte[] requestBytes = command.build().toByteArray();
+
+            // Execute via JNI - returns converted Java objects directly
+            // No need to wrap in Response since JNI already provides the final object
+            // Use binary or UTF-8 mode based on expected response type, not argument type
+            CompletableFuture<Object> jniFuture =
+                    expectUtf8Response
+                            ? coreClient.executeCommandAsync(requestBytes) // Force UTF-8 conversion
+                            : coreClient.executeBinaryCommandAsync(requestBytes); // Allow binary conversion
+
+            return jniFuture
+                    .thenApply(
+                            result -> {
+                                Response.Builder builder = Response.newBuilder();
+                                Object toStore = result;
+                                if (result == null) {
+                                    builder.setRespPointer(0L);
+                                } else if ("OK".equals(result)) {
+                                    builder.setConstantResponse(ConstantResponse.OK);
+                                } else {
+                                    if (result instanceof ByteBuffer) {
+                                        toStore = normalizeDirectBuffer((ByteBuffer) result, expectUtf8Response);
+                                    }
+                                    long objectId = JniResponseRegistry.storeObject(toStore);
+                                    builder.setRespPointer(objectId);
+                                }
+                                return responseHandler.apply(builder.build());
+                            })
+                    .exceptionally(this::exceptionHandler);
+        } catch (Exception e) {
+            var errorFuture = new CompletableFuture<T>();
+            errorFuture.completeExceptionally(e);
+            return errorFuture;
+        }
+    }
+
+    private Object normalizeDirectBuffer(ByteBuffer buffer, boolean expectUtf8Response) {
+        ByteBuffer dup = buffer.duplicate();
+        dup.rewind();
+        if (dup.remaining() == 0) {
+            return expectUtf8Response ? "" : glide.api.models.GlideString.gs(new byte[0]);
+        }
+        byte marker = dup.get();
+        dup.rewind();
+        if (marker == '*') {
+            // Serialized array/map (custom wire format)
+            return deserializeByteBufferArray(dup, expectUtf8Response);
+        } else if (marker == '%') {
+            return deserializeByteBufferMap(dup, expectUtf8Response);
+        }
+        // Bulk string bytes
+        byte[] bytes = new byte[dup.remaining()];
+        dup.get(bytes);
+        if (expectUtf8Response) {
+            return new String(bytes, java.nio.charset.StandardCharsets.UTF_8);
```

Comment:
Use StandardCharsets.UTF_8.decode(ByteBuffer) instead of copying to byte array temporary.